### PR TITLE
Add a real wall-clock time device

### PR DESCRIPTION
**Pull request by bshepherdson**
Comments welcome!

Is millisecond precision practical for the server's point of view?

### CODE REVIEW

**Comment by azertyfun** [comment]
- Is the time UTC or dependent on the server local time? Or is it some in-game arbitrary time?
- Why 12 hours clock? Without even an am/pm indicator, that is pretty ambiguous.
- Milliseconds seems weird ; the generic clock only ticks at 60 Hz, and the DCPU is clocked at 100 KHz. Expecting such a level of accuracy from a mere clock in-game would be like expecting accuracy to the nanosecond in a clock today.

**Comment by bshepherdson** [comment]
- I hadn't thought about time zones. Earth time zones don't make sense in a space travel game. Presumably there's some kind of universal time. I'd say for in-game it's that time. For other emulators, the host machine's local time seems fine to me.
- Unclear wording - it's in the range 0-23. I was just commenting that 12 was noon. I'll clarify that.
- By "wall clock" I don't mean that this is hardware driving an actual wall clock. It's just a euphemism for "absolute time and date". This is a computer's relatively fast view of time. That said, most hardware today has clock speeds in excess of 1 GHz (10000x faster than DCPU) and can track the time in microseconds (1000x more resolution than I'm proposing here). Reducing it to 100ths or 10ths of seconds would be fine with me, but I don't think a millisecond-accurate clock is out of place with the speed or supposed era of this hardware.

On another point, one of the reasons for this hardware is not the time or date exactly, but using the time and date (especially its high-resolution parts!) as entropy for a software PRNG. The lower the resolution of this clock, the harder that becomes.

**Comment by bshepherdson** [comment]
I've updated this. Now it's a backwards-compatible extension to the generic clock device, which now combines relative timer and absolute time/date functions.

I've also clarified that the time zone is not specified. Interpreting the hardware time is a software concern, just as it is on present-day hardware. It's recommended, but not required, that emulators use the host machine's local time.

File: CPU/clock.md [hunk 1]
```diff
@@ -6,24 +6,62 @@ Generic Clock
...
+All interrupts return the time in the same format:
+
+- `B`: Year (Earth year, CE/AD)
+- `C`: Month and date
+    - Month is high byte; 1 = January, 12 = December
+    - Date is days into that month, 1-31.
+- `X`: Hours and Minutes
+    - Hours are high byte; 0-23, where 0 is might, 12 is noon.
+    - Minutes are low byte; 0-59
+- 'Y': Seconds; 0-59
```
Comment (ike709) [inline CPU/clock.md]: `

**Review by paultech — APPROVED** [review]

**Comment by paultech** [comment]
This looks good to me. Waiting on @andyleap to weight on then we will merge and get it into the emulator/client.

File: CPU/clock.md [hunk 2]
```diff
@@ -6,24 +6,62 @@ Generic Clock
...
+- `Y`: Seconds; 0-59
+- `Z`: Milliseconds; 0-999
+
+Frame of Reference
+----
+The exact frame of reference for the absolute time is left unspecified.
+
+For DCPU-16 emulators, the host machine's local time is recommended.
+
+In a game universe, the clock might start uninitialized (all 0s) and need to be
```
Comment (andyleap) [inline CPU/clock.md]: Specs are supposed to be written as if the spec is for a "real thing", i.e. In Character.  Referencing emulators and game universe is a no-no

**Review by andyleap — CHANGES_REQUESTED** [review]
Strike the frame of reference bit, maybe replace it with something about "Your RTC may not come preset, please set time on first use" or something.

File: CPU/clock.md (hunk 2 above)
Comment (bshepherdson) [reply CPU/clock.md]: Fixed. Now in-character, with mention that batteries to hold the time without power are recommended, but not required.

I struck the bit about "frame of reference" - it was referring to time zones originally, but that doesn't make sense in character.

**Review by andyleap — APPROVED** [review]

**Comment by Yamakaky** [comment]
wrong indent, check the rendered version

File: CPU/clock.md [hunk 3]
```diff
@@ -6,24 +6,60 @@ Generic Clock
...
 
 Interrupt Commands
 ----
 
- - **0x0000**: SET_SPEED
+ - **0x0000**: `SET_SPEED`
  	- B: Clock is set to 60/B per second. Set B to zero to disable.
- - **0x0001**: GET_TICKS
+ - **0x0001**: `GET_TICKS`
  	- C: Number of ticks since last call in C
```
Comment (Yamakaky) [inline CPU/clock.md:26]: The phrase could be improved

File: CPU/clock.md [hunk 4]
```diff
@@ -6,24 +6,60 @@ Generic Clock
...
+
+This device is backwards compatible with the original version 1 clock.
+
 
 Interrupt Commands
 ----
 
- - **0x0000**: SET_SPEED
+ - **0x0000**: `SET_SPEED`
  	- B: Clock is set to 60/B per second. Set B to zero to disable.
```
Comment (Yamakaky) [inline CPU/clock.md:24]: Clock tick rate is set...

File: CPU/clock.md [hunk 5]
```diff
@@ -6,24 +6,60 @@ Generic Clock
...
+  - **0x0012**: `SET_REAL_TIME`
+	- Sets the absolute time, based on the format below.
+  - **0xffff**: `RESET`
+	- Resets to initial state, as though the DCPU was restarted. (Timer
+          disabled, elapsed run time reset to 0.)
+
+
+Time Format
+----
+All interrupts return the real time in the same format:
```
Comment (Yamakaky) [inline CPU/clock.md:42]: Not clear, at first I thought the spec didn't define how the time is returned

**Pull request merged**